api/handlers.go

```diff
@@ -882,8 +882,381 @@ func (a *App) handleUnsubscribe(c *gin.Context) {
+func (a *App) handleGetBookmarks(c *gin.Context) {
+	var r GetBookmarksRequest
+	if c.Bind(&r) != nil {
+		return
+	}
+
+	user := &models.User{}
+	if err := a.validateUser(c, user); err != nil {
+		err.Abort(c)
+		return
+	}
+
+	db := c.MustGet("MY_DB").(*sql.DB)
+
+	mods := []qm.QueryMod{models.BookmarkWhere.UserID.EQ(user.ID)}
+
+	total, err := models.Bookmarks(mods...).Count(db)
+
+	mods = append(mods, qm.Load(models.BookmarkRels.BookmarkFolders))
+	_, offset := appendListMods(&mods, r.ListRequest)
+	if int64(offset) >= total {
+		concludeRequest(c, new(GetBookmarksResponse), nil)
+		return
+	}
+	appendQueryFilter(&mods, r.QueryFilter, "name")
+	if len(r.FolderIDsFilter) > 0 {
+		mods = append(mods,
+			qm.InnerJoin("bookmark_folder bf ON id = bf.bookmark_id"),
+			qm.WhereIn("bf.folder_id in ?", utils.ConvertArgsInt64(r.FolderIDsFilter)...),
+		)
+	}
+
+	bookmarks, err := models.Bookmarks(mods...).All(db)
+	if err != nil {
+		errs.NewInternalError(pkgerr.WithStack(err)).Abort(c)
+		return
+	}
+
+	items := make([]*Bookmark, len(bookmarks))
+	for i, b := range bookmarks {
+		items[i] = makeBookmarkDTO(b)
+	}
+
+	resp := GetBookmarksResponse{
+		ListResponse: ListResponse{Total: total},
+		Items:        items,
+	}
+	concludeRequest(c, resp, err)
+}
+
+func (a *App) handleCreateBookmark(c *gin.Context) {
+	var r AddBookmarksRequest
+	if c.BindJSON(&r) != nil {
+		return
+	}
+
+	user := &models.User{}
+	if err := a.validateUser(c, user); err != nil {
+		err.Abort(c)
+		return
+	}
+
+	db := c.MustGet("MY_DB").(*sql.DB)
+	bookmark := &models.Bookmark{
+		UserID:     user.ID,
+		SourceUID:  r.SourceUID,
+		SourceType: r.SourceType,
+	}
+	if r.Name != "" {
+		bookmark.Name = null.StringFrom(r.Name)
+	}
+	if r.Data != nil {
+		data, err := json.Marshal(r.Data)
+		bookmark.Data = null.JSONFrom(data)
+		if err != nil {
+			errs.NewBadRequestError(err).Abort(c)
+			return
+		}
+	}
+	err := sqlutil.InTx(context.TODO(), db, func(tx *sql.Tx) error {
+		err := bookmark.Insert(tx, boil.Infer())
+		if err != nil {
+			return pkgerr.WithStack(err)
+		}
+		if r.FolderIDs != nil {
+			bbfs := make([]*models.BookmarkFolder, len(r.FolderIDs))
+			for i, id := range r.FolderIDs {
+				bbfs[i] = &models.BookmarkFolder{
+					FolderID: id,
+				}
+			}
+			if err := bookmark.AddBookmarkFolders(tx, true, bbfs...); err != nil {
+				return pkgerr.WithStack(err)
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		errs.NewInternalError(pkgerr.WithStack(err)).Abort(c)
+		return
+	}
+	resp := makeBookmarkDTO(bookmark)
+	resp.FolderIds = r.FolderIDs
+	concludeRequest(c, resp, err)
+}
+
+func (a *App) handleUpdateBookmark(c *gin.Context) {
+	id, err := strconv.ParseInt(c.Param("id"), 10, 0)
+	if err != nil {
+		errs.NewBadRequestError(pkgerr.Wrap(err, "id expects int64")).Abort(c)
+		return
+	}
+
+	var r UpdateBookmarkRequest
+	if c.BindJSON(&r) != nil {
+		return
+	}
+
+	user := &models.User{}
+	if err := a.validateUser(c, user); err != nil {
+		err.Abort(c)
+		return
+	}
+
```

Comment:
need to make sure the user owns this bookmark or this is a security breach

Reply:
fixed